### PR TITLE
Implemented support for filters and removable routes in ESP8266WebServer

### PR DESCRIPTION
**Pull request by ayushsharma82**
This PR implements filters and removable routes in ESP8266 arduino core, making it API compatible with recent changes to ESP32 WebServer API.

Reference (Recent additions to ESP32 WebServer API): 
- https://github.com/espressif/arduino-esp32/pull/9832
- https://github.com/espressif/arduino-esp32/pull/9842
- https://github.com/espressif/arduino-esp32/pull/9851

__Noteable Changes:__
1. Request Filters (`setFilter`)
2. Removable Routes (`removeRoute`, `removeHandler`, `_removeRequestHandler`)

### CODE REVIEW

**Comment by ayushsharma82** [comment]
@me-no-dev , Please review this as well.

**Comment by mcspr** [comment]
> @me-no-dev , Please review this as well.

Not an Espressif repo (maybe ping @d-a-v). Will merge based on esp32 review, though.
Also, please take a look at 'doc/' and WebServer's 'README.md' and make sure everything is up-to-date w/ web documentation related to WebServer

**Comment by ayushsharma82** [comment]
Oh. I thought me-no-dev is a maintainer in this repo too. 

I'll check documentation.

**Comment by me-no-dev** [comment]
> Oh. I thought me-no-dev is a maintainer in this repo too.

Long time ago :)

**Comment by ayushsharma82** [comment]
@mcspr , @d-a-v, 

ESP32 PRs are merged therefore this PR is ready for review & merger as well.

**Comment by ayushsharma82** [comment]
Any idea why LwIP CI tests are failing? Compiling on Arduino IDE locally doesn't show any error.

**Comment by mcspr** [comment]
>  Multiple libraries were found for "WiFi.h"
   Used: /home/runner/work/_temp/arduino_ide/libraries/WiFi
   Not used: /home/runner/Arduino/hardware/esp8266com/esp8266/libraries/ESP8266WiFi

File: libraries/ESP8266WebServer/examples/Filters/Filters.ino [hunk 1]
```diff
@@ -0,0 +1,101 @@
+#include <WiFi.h>
```
Comment (mcspr) [inline libraries/ESP8266WebServer/examples/Filters/Filters.ino:1]:
```suggestion
#include <ESP8266WiFi.h>
```
?

We do have wrapper headers, but for some reason IDE does not like them :/
https://github.com/esp8266/Arduino/blob/master/libraries/ESP8266WiFi/src/WiFi.h

Maybe order, maybe 1.x vs. 2.x

**Comment by ayushsharma82** [comment]
Great! Everything compiles now, what should we do about style test?

File: libraries/ESP8266WebServer/examples/Filters/Filters.ino [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  })
+  .setFilter(ON_STA_FILTER);
+
+  // This route will be accessible by AP clients only
+  server.on("/", [&]() {
+    digitalWrite(led, 1);
+    server.send(200, "text/plain", "Hi!, This route is accessible for AP clients only");
+    digitalWrite(led, 0);
+  })
+  .setFilter(ON_AP_FILTER);
```
Comment (mcspr) [inline libraries/ESP8266WebServer/examples/Filters/Filters.ino:87]: https://github.com/esp8266/Arduino/actions/runs/9516245776/job/26232190207#step:4:400
./tests/restyle.sh runs clang-format-15 on everything, chaining makes extra indentation

lose `[&]` and maybe do something explicit with 'on' return value?
```cpp
auto& onlyAp = ...;
onlyAp.setFilter(ON_AP_FILTER);
```

Comment (ayushsharma82) [reply libraries/ESP8266WebServer/examples/Filters/Filters.ino:87]: I tried running `./tests/restyle.sh` on my PC, odd enough that did nothing. Do we have to install anything for it to run?

_**Regarding Example:**
I think it's better if we keep it same as ESP32 Filters.ino example, that way people are not confused._

Comment (mcspr) [reply libraries/ESP8266WebServer/examples/Filters/Filters.ino:87]: `clang-format-15` in $PATH / env CLANG_FORMAT=path-to-clang-format ./tests/restyle.sh
(e.g. https://github.com/muttleyxd/clang-tools-static-binaries)

Comment (ayushsharma82) [reply libraries/ESP8266WebServer/examples/Filters/Filters.ino]: Done

Comment (mcspr) [reply libraries/ESP8266WebServer/examples/Filters/Filters.ino]: > I think it's better if we keep it same as ESP32 Filters.ino example, that way people are not confused.

Right, but it definitely got me confused
- `[&]` meant to capture by reference, but server callback should not do that. if you do it in setup() and capture value from there, later callback in loop() references invalid memory location
- nothing is actually captured, don't wanna do that anyway

Specifically for us here and clang-format, without chaining - there are less spaces.

**Pull request merged**